File: core/tests/python/conformance_lib/codec/well_formed.py
"""A whole-body well-formedness walk, run before anything is interpreted (#641).

The clean-room twin of `core/src/cbor/well_formed.rs`.  `py_decode_record` runs
it first, so a body that is not well-formed CBOR is reported as that, and a
well-formed body carrying a tag or a float as crypto-design §6.2 rule 4,
before any key is read -- the order `record::decode` reports them in.

WHAT IT CHECKS -- RFC 8949 well-formedness plus vault-format §4.2's
precondition list, then rule 4:
  * a truncated head, argument or payload, and an indefinite item with no break;
  * reserved additional-info 28-30, the indefinite form on majors 0, 1 and 6,
    and a stray break;
  * an indefinite-string chunk that is not a definite string of the same major
    (RFC 8949 §3.2.3) -- which rejects a NESTED indefinite chunk;
  * text that is not valid UTF-8, per string and per chunk (a sequence split
    across two chunks is invalid; ciborium holds the same, measured);
  * a major-7 simple value other than false/true/null;
  * a chain of arrays, maps and tags nested past crypto-design §6.2 rule 6's
    limit of 256, as `NestingTooDeep` (see ITERATIVE below);
  * then any tag (bignum tags 2 and 3 included) or float, as rule 4.

PRECEDENCE.  A well-formedness fault anywhere in the item outranks a rule-4
fault anywhere: the first tag or float is remembered and raised only once the
whole item has proven well-formed.

ITERATIVE on purpose, and bounded by crypto-design §6.2 rule 6 (#667).
`scanner._scan_item` recurses, so a deeply nested body used to raise
`RecursionError` -- a harness failure, not a verdict.  This walk keeps an
explicit stack, and the stack is the depth count: a head that would open level
257 raises `NestingTooDeep` at once, like every well-formedness fault, so no
recursive phase after it ever sees more than 256 levels.  `reject_excessive_nesting`
is the same traversal with the content checks off, for the decoders that have
no `walk_body` of their own.

SCOPE.  The first item only.  Trailing bytes are the caller's to judge, and
`py_decode_record` judges them LAST, where `record::decode` meets them: its
parse performs no EOF check.
"""

from __future__ import annotations

from dataclasses import dataclass

from conformance_lib.codec.cbor_faults import MalformedCbor, NestingTooDeep, require_false_true_or_null, require_room_for_another_level, require_utf8
from conformance_lib.codec.scanner import CBOR_AI_INDEFINITE, CBOR_BREAK, NonCanonicalItem, _decode_head, _reject_rule4_head

MAJOR_UINT, MAJOR_NINT, MAJOR_BYTES, MAJOR_TEXT, MAJOR_ARRAY, MAJOR_MAP, MAJOR_TAG, MAJOR_SIMPLE = range(8)
# A map's items are keys and values.
ITEMS_PER_MAP_ENTRY = 2


@dataclass
class _Frame:
    """An open container.  A definite one counts the items it still needs (a
    map counts keys and values; a tag needs one); an indefinite map tracks
    whether a key is waiting for its value."""

    definite_left: int | None
    is_map: bool = False
    mid_entry: bool = False


def _rule4_at(major: int, ai: int, pos: int) -> NonCanonicalItem | None:
    try:
        _reject_rule4_head(major, ai, pos)
    except NonCanonicalItem as exc:
        return exc
    return None


def _payload_end(buf: bytes, head_at: int, start: int, length: int, text: bool) -> int:
    end = start + length
    if end > len(buf):
        raise MalformedCbor(f"string length {length} overruns buffer at offset {head_at}")
    if text:
        require_utf8(buf, start, end, head_at)
    return end


def _string_end(buf: bytes, pos: int, major: int, arg: int | None, head: int, check_utf8: bool) -> int:
    text = major == MAJOR_TEXT and check_utf8
    if arg is not None:
        return _payload_end(buf, pos, pos + head, arg, text)
    at = pos + head
    while True:
        if at >= len(buf):
            raise MalformedCbor("unterminated indefinite-length string")
        if buf[at] == CBOR_BREAK:
            return at + 1
        chunk_major, _, chunk_arg, chunk_head = _decode_head(buf, at)
        if chunk_major != major or chunk_arg is None:
            raise MalformedCbor(f"bad chunk in indefinite-length string at {at}")
        at = _payload_end(buf, at, at + chunk_head, chunk_arg, text)


def _close_finished(buf: bytes, pos: int, stack: list[_Frame]) -> int:
    while stack:
        top = stack[-1]
        if top.definite_left == 0:
            stack.pop()
            continue
        at_break = pos < len(buf) and buf[pos] == CBOR_BREAK
        if top.definite_left is None and at_break:
            if top.is_map and top.mid_entry:
                raise MalformedCbor(
                    f"indefinite-length map ends between a key and its value at offset {pos}"
                )
            stack.pop()
            pos += 1
            continue
        break
    return pos


def _count_one_item(stack: list[_Frame]) -> None:
    if not stack:
        return
    top = stack[-1]
    if top.definite_left is not None:
        top.definite_left -= 1
    elif top.is_map:
        top.mid_entry = not top.mid_entry


def _walk(buf: bytes, pos: int, *, check_content: bool) -> int:
    """The one traversal both entry points share: item boundaries, crypto-design
    §6.2 rule 6 always, and -- when `check_content` -- UTF-8, simple values and
    rule 4.  The depth check lives here once, so it cannot drift between them."""
    stack: list[_Frame] = []
    first_rule4: NonCanonicalItem | None = None
    started = False
    while True:
        pos = _close_finished(buf, pos, stack)
        if started and not stack:
            if first_rule4 is not None:
                raise first_rule4
            return pos
        started = True
        _count_one_item(stack)
        major, ai, arg, head = _decode_head(buf, pos)
        if major in (MAJOR_UINT, MAJOR_NINT):
            pos += head
        elif major in (MAJOR_BYTES, MAJOR_TEXT):
            pos = _string_end(buf, pos, major, arg, head, check_utf8=check_content)
        elif major in (MAJOR_ARRAY, MAJOR_MAP):
            require_room_for_another_level(len(stack), pos)
            is_map = major == MAJOR_MAP
            left = None if arg is None else arg * (ITEMS_PER_MAP_ENTRY if is_map else 1)
            stack.append(_Frame(definite_left=left, is_map=is_map))
            pos += head
        elif major == MAJOR_TAG:
            require_room_for_another_level(len(stack), pos)
            if check_content:
                first_rule4 = first_rule4 or _rule4_at(major, ai, pos)
            stack.append(_Frame(definite_left=1))
            pos += head
        else:
            if ai == CBOR_AI_INDEFINITE:
                raise MalformedCbor(f"unexpected break at offset {pos}")
            if check_content:
                rule4 = _rule4_at(major, ai, pos)
                if rule4 is None:
                    require_false_true_or_null(ai, pos)
                first_rule4 = first_rule4 or rule4
            pos += head


def walk_body(buf: bytes, pos: int = 0) -> int:
    """Walk the CBOR item at `pos`; return the offset one past it.

    Raises `MalformedCbor` for a well-formedness fault anywhere in the item --
    `NestingTooDeep`, a subclass, for a level past crypto-design §6.2 rule 6,
    at once -- else `NonCanonicalItem` (rule 4) for the first tag or float.
    """
    return _walk(buf, pos, check_content=True)


def reject_excessive_nesting(buf: bytes, *, later_phases_scan_in_byte_order: bool) -> None:
    """crypto-design §6.2 rule 6 over a whole document, and nothing else (#667).

    The first statement of every `codec/` decoder that has no `walk_body` of its
    own (the manifest, the contact card, the trash entry), so a document nested
    past the limit is refused before any RECURSIVE phase runs -- which is what
    turned a 995-level manifest into a `RecursionError` harness failure.

    It walks item boundaries only.  It reports no content-level fault (invalid
    UTF-8, a disallowed simple value, a tag or float as rule 4), since none of
    those moves a boundary; a tag still counts as a level.

    At a STRUCTURAL fault it cannot walk past -- a truncated head, an overrun, a
    bad chunk, a stray break -- what it does depends on the caller, which must
    say, because the answer is a property of the caller's later phases:

      * `later_phases_scan_in_byte_order=True` (the manifest, whose later phases
        are `scanner._scan_item`): it returns, leaving the fault to those phases
        to report as they always have.  They scan in byte order and raise at
        the same byte, so none of them can nest past 256 levels first.
      * `later_phases_scan_in_byte_order=False` (the contact card and the trash
        entry, whose next phase is `cbor2.loads`): it raises the fault.  cbor2
        is NOT a byte-order well-formedness check -- it accepts a stray break
        inside a definite array, returning a sentinel object -- so a body with
        a break ahead of a 300-level chain used to pass this pass silently and
        be rejected by the ENCODER failing on that sentinel (PR #684 review).
    """
    try:
        _walk(buf, 0, check_content=False)
    except NestingTooDeep:
        raise
    except MalformedCbor:
        if not later_phases_scan_in_byte_order:
            raise

File: core/src/cbor/well_formed.rs
//! A byte-level walk over the first CBOR item: well-formedness first, then
//! crypto-design §6.2 rule 4 (#641).
//!
//! **Why it exists.** ciborium's `Value` reader lets four forms through that
//! this format rejects, and each changes WHICH error a record reports, which
//! `core/tests/differential_replay.rs` compares against `conformance.py`. They
//! break different rules, so name the rule per form:
//!
//! - `undefined` (`0xf7`), read as `null`. RFC 8949 calls it well-formed;
//!   `docs/vault-format.md` §4.2's well-formedness precondition excludes it
//!   ("a major-7 value outside `false`/`true`/`null`").
//! - The two-byte simple forms `f8 14`..`f8 17`, read as false, true, null and
//!   undefined. RFC 8949 §3.3 makes those encodings not well-formed.
//! - A nested indefinite-length string chunk, which RFC 8949 §3.2.3 forbids.
//! - A bignum (tag 2 or 3) whose value fits in 64 bits, turned into an
//!   integer, so a later parsed-tree rule-4 walk never sees the tag. It is
//!   well-formed; it breaks crypto-design §6.2 rule 4. A wider bignum stays a
//!   `Value::Tag` in ciborium 0.2.2, which that walk does see.
//!
//! The corpus measured the first, third and fourth; the two-byte simple form
//! was found in ciborium's source in the PR #673 review. `record_walk_tests`
//! pins all four, and the wide bignum. None changes whether a record is
//! accepted — none is canonical, so the re-encode comparison rejects them all
//! — and walking the bytes before ciborium reports what the bytes actually are.
//!
//! **What it checks.** RFC 8949 well-formedness plus §4.2's precondition list:
//! a truncated head, argument or payload; reserved additional-info 28-30; the
//! indefinite form on majors 0, 1 and 6; a break outside an indefinite
//! container; an indefinite-string chunk that is not a definite string of the
//! same major (§3.2.3); text that is not valid UTF-8, per string and per chunk;
//! a major-7 simple value other than false/true/null; a chain of arrays, maps
//! and tags nested past crypto-design §6.2 rule 6's limit. Then rule 4: any tag
//! (bignum tags included) and any float.
//!
//! **Precedence.** A well-formedness fault anywhere in the item outranks a
//! rule-4 fault anywhere: the first tag or float is remembered and reported only
//! once the whole item has proven well-formed.
//!
//! **Iterative, and bounded by crypto-design §6.2 rule 6.** An explicit stack,
//! no recursion. The stack is also the depth count: a head that would open a
//! level past [`V1_MAX_NESTING_DEPTH`] (arrays, maps and tags alike; a scalar
//! is not a level) is `Malformed` with kind `RecursionLimit`, reported at once
//! like every well-formedness fault. It is the limit `ciborium`'s parse
//! applies afterwards, so on the record path the walk now answers before
//! `ciborium` can (#667).
//!
//! **Pure.** It reads a byte slice and allocates only its container stack;
//! nothing it holds is a copy of a payload.
//!
//! Python's twin is `conformance_lib/codec/well_formed.py`'s `walk_body`.

use crate::cbor::{CborErrorKind, CborFault, V1_MAX_NESTING_DEPTH};

/// Why [`walk_first_item`] stopped.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum WalkFault {
    /// Not well-formed CBOR. `Syntax` at the offending head. `Io` when the
    /// input ended first: at the head whose argument or payload runs past the
    /// end, or at the end itself when a head or a break was still needed.
    Malformed(CborFault),
    /// §6.2 rule 4: a tag, at `offset`.
    Tag { offset: usize },
    /// §6.2 rule 4: a float, at `offset`.
    Float { offset: usize },
}

const MAJOR_SHIFT: u8 = 5;
const AI_MASK: u8 = 0x1f;
const MAJOR_UINT: u8 = 0;
const MAJOR_NINT: u8 = 1;
const MAJOR_BYTES: u8 = 2;
const MAJOR_TEXT: u8 = 3;
const MAJOR_ARRAY: u8 = 4;
const MAJOR_MAP: u8 = 5;
const MAJOR_TAG: u8 = 6;
/// The largest additional-info value that is itself the argument.
const AI_DIRECT_MAX: u8 = 23;
const AI_ONE_BYTE: u8 = 24;
const AI_TWO_BYTES: u8 = 25;
const AI_FOUR_BYTES: u8 = 26;
const AI_EIGHT_BYTES: u8 = 27;
const AI_INDEFINITE: u8 = 31;
const SIMPLE_FALSE: u8 = 20;
const SIMPLE_TRUE: u8 = 21;
const SIMPLE_NULL: u8 = 22;
/// For major 7, additional-info 25/26/27 is a half/single/double float.
const AI_FLOAT16: u8 = AI_TWO_BYTES;
const AI_FLOAT32: u8 = AI_FOUR_BYTES;
const AI_FLOAT64: u8 = AI_EIGHT_BYTES;
const BREAK: u8 = 0xff;
const BITS_PER_BYTE: u32 = 8;
/// A map's items are keys and values.
const ITEMS_PER_MAP_ENTRY: u64 = 2;

struct Head {
    major: u8,
    ai: u8,
    /// `None` for the indefinite form.
    arg: Option<u64>,
    len: usize,
}

/// Which rule-4 fault was seen first.
///
/// Its own type rather than a [`WalkFault`], which can also hold `Malformed`:
/// the walk parks this until the whole item has proven well-formed, and a
/// well-formedness fault parked here would be reported late, silently
/// breaking the precedence the module doc states (PR #673 review).
#[derive(Clone, Copy)]
enum Rule4 {
    Tag,
    Float,
}

/// An open container.
enum Frame {
    /// A definite array or map, or a tag's single content item, with `left`
    /// items still to read (a map counts keys and values).
    Definite {
        left: u64,
    },
    IndefiniteArray,
    /// `mid_entry` is true after a key and before its value.
    IndefiniteMap {
        mid_entry: bool,
    },
}

fn end_of_input(offset: usize) -> CborFault {
    CborFault {
        kind: CborErrorKind::Io,
        offset: Some(offset),
    }
}

fn syntax(offset: usize) -> CborFault {
    CborFault {
        kind: CborErrorKind::Syntax,
        offset: Some(offset),
    }
}

fn read_head(bytes: &[u8], at: usize) -> Result<Head, CborFault> {
    let Some(&initial) = bytes.get(at) else {
        return Err(end_of_input(at));
    };
    let major = initial >> MAJOR_SHIFT;
    let ai = initial & AI_MASK;
    let arg_len = match ai {
        0..=AI_DIRECT_MAX => {
            return Ok(Head {
                major,
                ai,
                arg: Some(u64::from(ai)),
                len: 1,
            })
        }
        AI_ONE_BYTE => 1,
        AI_TWO_BYTES => 2,
        AI_FOUR_BYTES => 4,
        AI_EIGHT_BYTES => 8,
        AI_INDEFINITE => {
            // RFC 8949 §3.2: valid for strings, arrays and maps, and for
            // major 7 it is the break code. Never for integers or tags.
            return if matches!(major, MAJOR_UINT | MAJOR_NINT | MAJOR_TAG) {
                Err(syntax(at))
            } else {
                Ok(Head {
                    major,
                    ai,
                    arg: None,
                    len: 1,
                })
            };
        }
        _ => return Err(syntax(at)),
    };
    let Some(arg_bytes) = bytes.get(at + 1..at + 1 + arg_len) else {
        return Err(end_of_input(at));
    };
    let arg = arg_bytes
        .iter()
        .fold(0u64, |acc, byte| (acc << BITS_PER_BYTE) | u64::from(*byte));
    Ok(Head {
        major,
        ai,
        arg: Some(arg),
        len: 1 + arg_len,
    })
}

/// The offset one past a string payload of `len` bytes starting at `start`,
/// checking UTF-8 for text. Faults are reported at the string's head.
fn payload_end(
    bytes: &[u8],
    head_at: usize,
    start: usize,
    len: u64,
    text: bool,
) -> Result<usize, CborFault> {
    let end = usize::try_from(len)
        .ok()
        .and_then(|len| start.checked_add(len));
    let Some(content) = end.and_then(|end| bytes.get(start..end)) else {
        return Err(end_of_input(head_at));
    };
    if text && std::str::from_utf8(content).is_err() {
        return Err(syntax(head_at));
    }
    Ok(start + content.len())
}

/// The offset one past the string whose head is at `head_at`.
fn string_end(bytes: &[u8], head_at: usize, head: &Head) -> Result<usize, CborFault> {
    let text = head.major == MAJOR_TEXT;
    let start = head_at + head.len;
    let Some(len) = head.arg else {
        return chunks_end(bytes, start, head.major, text);
    };
    payload_end(bytes, head_at, start, len, text)
}

/// The offset one past the break that ends an indefinite string's chunks.
fn chunks_end(bytes: &[u8], mut at: usize, major: u8, text: bool) -> Result<usize, CborFault> {
    loop {
        match bytes.get(at) {
            None => return Err(end_of_input(at)),
            Some(&BREAK) => return Ok(at + 1),
            Some(_) => {
                let chunk = read_head(bytes, at)?;
                // §3.2.3: every chunk is a DEFINITE string of the same major.
                // Its payload is read directly rather than through
                // `string_end`, so a chunk can never re-enter this loop.
                let Some(len) = chunk.arg.filter(|_| chunk.major == major) else {
                    return Err(syntax(at));
                };
                at = payload_end(bytes, at, at + chunk.len, len, text)?;
            }
        }
    }
}

/// Close every container whose items are complete, or whose break is next.
fn close_finished_containers(
    bytes: &[u8],
    pos: &mut usize,
    stack: &mut Vec<Frame>,
) -> Result<(), CborFault> {
    loop {
        let at_break = bytes.get(*pos) == Some(&BREAK);
        match stack.last() {
            Some(Frame::Definite { left: 0 }) => {
                stack.pop();
            }
            Some(Frame::IndefiniteArray) if at_break => {
                stack.pop();
                *pos += 1;
            }
            Some(Frame::IndefiniteMap { mid_entry }) if at_break => {
                if *mid_entry {
                    return Err(syntax(*pos));
                }
                stack.pop();
                *pos += 1;
            }
            _ => return Ok(()),
        }
    }
}

/// Account for one item about to be read inside `parent`.
fn count_one_item(parent: Option<&mut Frame>) {
    match parent {
        Some(Frame::Definite { left }) => *left -= 1,
        Some(Frame::IndefiniteMap { mid_entry }) => *mid_entry = !*mid_entry,
        Some(Frame::IndefiniteArray) | None => {}
    }
}

fn open_container(head: &Head) -> Frame {
    match (head.major, head.arg) {
        (MAJOR_ARRAY, Some(count)) => Frame::Definite { left: count },
        (_, Some(count)) => Frame::Definite {
            left: count.saturating_mul(ITEMS_PER_MAP_ENTRY),
        },
        (MAJOR_ARRAY, None) => Frame::IndefiniteArray,
        (_, None) => Frame::IndefiniteMap { mid_entry: false },
    }
}

/// Open one more nesting level for the head at `head_at`, refusing the one
/// past crypto-design §6.2 rule 6. Returned at once, like every
/// well-formedness fault, so it outranks a rule-4 fault already remembered —
/// vault-format §4.2's precondition.
fn open_level(stack: &mut Vec<Frame>, head_at: usize, frame: Frame) -> Result<(), WalkFault> {
    if stack.len() >= V1_MAX_NESTING_DEPTH {
        return Err(WalkFault::Malformed(CborFault {
            kind: CborErrorKind::RecursionLimit,
            offset: Some(head_at),
        }));
    }
    stack.push(frame);
    Ok(())
}

/// Walk the first CBOR item in `bytes`; return the offset one past it.
///
/// See the module doc for what is checked and in what precedence.
pub(crate) fn walk_first_item(bytes: &[u8]) -> Result<usize, WalkFault> {
    let mut pos = 0usize;
    let mut stack: Vec<Frame> = Vec::new();
    let mut first_rule4: Option<(Rule4, usize)> = None;
    let mut started = false;
    loop {
        close_finished_containers(bytes, &mut pos, &mut stack).map_err(WalkFault::Malformed)?;
        if started && stack.is_empty() {
            return match first_rule4 {
                Some((Rule4::Tag, offset)) => Err(WalkFault::Tag { offset }),
                Some((Rule4::Float, offset)) => Err(WalkFault::Float { offset }),
                None => Ok(pos),
            };
        }
        started = true;
        count_one_item(stack.last_mut());
        let head = read_head(bytes, pos).map_err(WalkFault::Malformed)?;
        match head.major {
            MAJOR_UINT | MAJOR_NINT => pos += head.len,
            MAJOR_BYTES | MAJOR_TEXT => {
                pos = string_end(bytes, pos, &head).map_err(WalkFault::Malformed)?;
            }
            MAJOR_ARRAY | MAJOR_MAP => {
                open_level(&mut stack, pos, open_container(&head))?;
                pos += head.len;
            }
            MAJOR_TAG => {
                open_level(&mut stack, pos, Frame::Definite { left: 1 })?;
                first_rule4.get_or_insert((Rule4::Tag, pos));
                pos += head.len;
            }
            _ => {
                match head.ai {
                    SIMPLE_FALSE | SIMPLE_TRUE | SIMPLE_NULL => {}
                    AI_FLOAT16 | AI_FLOAT32 | AI_FLOAT64 => {
                        first_rule4.get_or_insert((Rule4::Float, pos));
                    }
                    // `undefined`, every unassigned simple value, the one-byte
                    // simple form, and a break outside an indefinite container.
                    _ => return Err(WalkFault::Malformed(syntax(pos))),
                }
                pos += head.len;
            }
        }
    }
}

#[cfg(test)]
mod tests;
